contracts/cw-distributor/src/contract.rs

```diff
@@ -0,0 +1,130 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{to_binary, CosmosMsg, Uint128, WasmMsg};
+
+use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
+use cw2::set_contract_version;
+use stake_cw20::msg::ReceiveMsg;
+
+use crate::error::ContractError;
+use crate::msg::{ConfigResponse, ExecuteMsg, InstantiateMsg, LastPaymentBlockResponse, QueryMsg};
+use crate::state::{Config, CONFIG, LAST_PAYMENT_BLOCK};
+
+const CONTRACT_NAME: &str = "crates.io:cw-distributor";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let owner = deps.api.addr_validate(&msg.owner)?;
+    let recipient = deps.api.addr_validate(&msg.recipient)?;
+
+    let reward_token = deps.api.addr_validate(&msg.reward_token)?;
+    let config = Config {
+        owner,
+        recipient,
+        reward_token,
+        reward_rate: msg.reward_rate,
+    };
+    CONFIG.save(deps.storage, &config)?;
+
+    // Initialize last payment block
+    LAST_PAYMENT_BLOCK.save(deps.storage, &env.block.height)?;
+
+    Ok(Response::new().add_attribute("action", "instantiate"))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateConfig {
+            owner,
+            recipient,
+            reward_rate,
+            reward_token,
+        } => execute_update_config(deps, info, owner, recipient, reward_rate, reward_token),
+        ExecuteMsg::Distribute {} => execute_distribute(deps, env),
+    }
+}
+
+pub fn execute_update_config(
+    deps: DepsMut,
+    info: MessageInfo,
+    owner: String,
+    recipient: String,
+    reward_rate: Uint128,
+    reward_token: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.owner != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let owner = deps.api.addr_validate(&owner)?;
+    let recipient = deps.api.addr_validate(&recipient)?;
+    let reward_token = deps.api.addr_validate(&reward_token)?;
+
+    let config = Config {
+        owner,
+        recipient,
+        reward_token,
+        reward_rate,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::default())
+}
+
+pub fn execute_distribute(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
```

Comment:
This function will lock tokens in the contract is `config.reward_rate * block_diff` is greater than the contract's token balance. This is likely to happen when the contract is coming to the end of its reward period. How would you feel about making the amount to send the min of remaining balance and `amount`?

Reply:
Likely want a test for this situation as well. :)

Reply:
Great suggestion. I'll add that

Reply:
yup, had same thought and +1 here

Reply:
Done